### ISSUE TITLE
V3 System.InvalidCastException of the Permissions field in the Identity module (use of Elsa.EntityFrameworkCore.SqlServer)

### ISSUE THREAD

**Issue by dao74** [elsa-workflows/elsa-core#3881]
Loading / mapping the Role entity cause an InvalidCastException, the stack trace is below:

```System.InvalidCastException: Unable to cast object of type 'System.Collections.Generic.List`1[System.String]' to type 'System.String'.
   at Elsa.EntityFrameworkCore.Modules.Identity.EFCoreRoleStore.Load(IdentityElsaDbContext dbContext, Role role) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Modules\Identity\RoleStore.cs:line 67
   at Elsa.EntityFrameworkCore.Common.Store`2.<>c__DisplayClass21_0.<QueryAsync>b__0(TEntity x) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Common\Store.cs:line 249
   at System.Linq.Enumerable.SelectListIterator`2.ToList()
   at Elsa.EntityFrameworkCore.Common.Store`2.QueryAsync(Func`2 query, Func`3 onLoading, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Common.Store`2.QueryAsync(Func`2 query, Func`3 onLoading, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Common\Store.cs:line 251
   at Open.Linq.AsyncExtensions.Extensions.ToList[TSource](Task`1 source)
   at Elsa.EntityFrameworkCore.Modules.Identity.EFCoreRoleStore.FindManyAsync(RoleFilter filter, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Modules\Identity\RoleStore.cs:line 52
   at Elsa.Identity.Providers.StoreBasedRoleProvider.FindManyAsync(RoleFilter filter, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Providers\StoreBasedRoleProvider.cs:line 27
   at Elsa.Extensions.RoleProviderExtensions.FindByIdsAsync(IRoleProvider roleProvider, IEnumerable`1 ids, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Extensions\RoleProviderExtensions.cs:line 27
   at Elsa.Identity.Services.DefaultAccessTokenIssuer.IssueTokensAsync(User user, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Services\DefaultAccessTokenIssuer.cs:line 34
   at Elsa.Identity.Endpoints.Login.Login.ExecuteAsync(Request request, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Endpoints\Login\Endpoint.cs:line 33
   at FastEndpoints.Endpoint`2.ExecAsync(CancellationToken ct)
   at FastEndpoints.Endpoint`2.ExecAsync(CancellationToken ct)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Elsa.Http.Middleware.WorkflowsMiddleware.InvokeAsync(HttpContext httpContext) in C:\Temp\elsa-core-3\src\modules\Elsa.Http\Middleware\WorkflowsMiddleware.cs:line 85
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)```

Similar behaviour in the save part:
```System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Collections.Generic.ICollection`1[System.String]'.
   at Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer`1.Equals(Object left, Object right)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.AreEqual(Object value, Object otherValue, IProperty property)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetProperty(IPropertyBase propertyBase, Object value, Boolean isMaterialization, Boolean setModified, Boolean isCascadeDelete, CurrentValueType valueType)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetProperty(IPropertyBase propertyBase, Object value, Boolean isMaterialization, Boolean setModified, Boolean isCascadeDelete)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.set_Item(IPropertyBase propertyBase, Object value)
   at Microsoft.EntityFrameworkCore.ChangeTracking.MemberEntry.set_CurrentValue(Object value)
   at Elsa.EntityFrameworkCore.Modules.Identity.EFCoreRoleStore.Save(IdentityElsaDbContext dbContext, Role role) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Modules\Identity\RoleStore.cs:line 59
   at Elsa.EntityFrameworkCore.Modules.Identity.EFCoreRoleStore.Load(IdentityElsaDbContext dbContext, Role role) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Modules\Identity\RoleStore.cs:line 68
   at Elsa.EntityFrameworkCore.Common.Store`2.<>c__DisplayClass21_0.<QueryAsync>b__0(TEntity x) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Common\Store.cs:line 249
   at System.Linq.Enumerable.SelectListIterator`2.ToList()
   at Elsa.EntityFrameworkCore.Common.Store`2.QueryAsync(Func`2 query, Func`3 onLoading, CancellationToken cancellationToken)
   at Elsa.EntityFrameworkCore.Common.Store`2.QueryAsync(Func`2 query, Func`3 onLoading, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Common\Store.cs:line 251
   at Open.Linq.AsyncExtensions.Extensions.ToList[TSource](Task`1 source)
   at Elsa.EntityFrameworkCore.Modules.Identity.EFCoreRoleStore.FindManyAsync(RoleFilter filter, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.EntityFrameworkCore\Modules\Identity\RoleStore.cs:line 53
   at Elsa.Identity.Providers.StoreBasedRoleProvider.FindManyAsync(RoleFilter filter, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Providers\StoreBasedRoleProvider.cs:line 27
   at Elsa.Extensions.RoleProviderExtensions.FindByIdsAsync(IRoleProvider roleProvider, IEnumerable`1 ids, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Extensions\RoleProviderExtensions.cs:line 27
   at Elsa.Identity.Services.DefaultAccessTokenIssuer.IssueTokensAsync(User user, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Services\DefaultAccessTokenIssuer.cs:line 34
   at Elsa.Identity.Endpoints.Login.Login.ExecuteAsync(Request request, CancellationToken cancellationToken) in C:\Temp\elsa-core-3\src\modules\Elsa.Identity\Endpoints\Login\Endpoint.cs:line 33
   at FastEndpoints.Endpoint`2.ExecAsync(CancellationToken ct)
   at FastEndpoints.Endpoint`2.ExecAsync(CancellationToken ct)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Elsa.Http.Middleware.WorkflowsMiddleware.InvokeAsync(HttpContext httpContext) in C:\Temp\elsa-core-3\src\modules\Elsa.Http\Middleware\WorkflowsMiddleware.cs:line 85
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)```

I can't figure out why a cast is necessary (first from list to a string and then back to a list), since the Permissions field was filled correctly as a collection/list from the database. Code block as follow:

```
    private static Role? Load(IdentityElsaDbContext dbContext, Role? role)
    {
        if (role == null)
            return null; 

        var permissionsJson = (string?)dbContext.Entry(role).Property("Permissions").CurrentValue;

        if (string.IsNullOrWhiteSpace(permissionsJson)) 
            return role;
        
        role.Permissions = JsonSerializer.Deserialize<ICollection<string>>(permissionsJson)!;
        return role;
    }

**Comment by sfmskywalker** [comment]
You were absolutely correct - none of this serialization business was necessary, since the `Role` entity was already configured with custom converter for its `Permissions` property.

I'll push a fix shortly, as well as an extra API endpoint to create roles.

**Comment by sfmskywalker** [comment]
Fixed as of 39c34b5a1e64beeac4f73b0e4e4af69d0f86a7ce
